lms/djangoapps/lti_provider/tasks.py

```diff
@@ -48,51 +47,51 @@ def score_changed_handler(sender, **kwargs):  # pylint: disable=unused-argument
+    assignments = outcomes.get_assignments_for_problem(
+        problem_descriptor, user_id, course_key
     )
 
-    # Calculate the user's score, on a scale of 0.0 - 1.0.
-    score = float(points_earned) / float(points_possible)
+    # Dictionary to hold the scores for each assignment. We already know the
+    # score for the problem that triggered this task.
+    scores = {usage_key: float(points_earned) / float(points_possible)}
 
-    # There may be zero or more assignment records. We would expect for there
-    # to be zero if the user/course/usage combination does not relate to a
-    # previous graded LTI launch. This can happen if an LTI consumer embeds some
-    # gradable content in a context that doesn't require a score (maybe by
-    # including an exercise as a sample that students may complete but don't
-    # count towards their grade).
-    # There could be more than one GradedAssignment record if the same content
-    # is embedded more than once in a single course. This would be a strange
-    # course design on the consumer's part, but we handle it by sending update
-    # messages for all launches of the content.
-    for assignment in assignments:
-        xml = lti_provider.outcomes.generate_replace_result_xml(
-            assignment.lis_result_sourcedid, score
+    # To find the score for a composite block (such as a vertical) we need to
+    # calculate the scores for the course. This is expensive, so we try to
+    # short-circuit it if we can. In the case where the graded launch contained
+    # only a single problem, and that problem is not a part of any other graded
+    # assignment, we can just use the score passed to this method by the signal
+    # handler.
+    if len(assignments) != 1 or problem_descriptor not in assignments:
+        # This is not the short-circuit case, so we need to calculate the score
+        # for at least one composite block. We do this by calculating the scores
+        # for all problems in the course, and then combining them to make up the
+        # composite score.
+        location_to_score = outcomes.get_scores_for_locations(
+            user_id, course_key
         )
-        try:
-            response = lti_provider.outcomes.sign_and_send_replace_result(assignment, xml)
-        except RequestException:
-            # failed to send result. 'response' is None, so more detail will be
-            # logged at the end of the method.
-            response = None
-            log.exception("Outcome Service: Error when sending result.")
+        for descriptor in assignments:
+            # A location could be part of multiple graded assignments
+            if descriptor.location not in scores:
+                earned, possible = outcomes.calculate_score(
+                    descriptor, location_to_score
+                )
+                if possible == 0:
+                    score = 0
+                else:
+                    score = earned / possible
```

Comment:
May be safer to explicitly cast `float(earned) / possible` to make sure there's no chance of integer division.